design/volume-group-snapshot.md

```diff
@@ -0,0 +1,428 @@
+# Add Support for VolumeGroupSnapshots
+
+This proposal outlines the design and implementation plan for incorporating VolumeGroupSnapshot support into Velero. The enhancement will allow Velero to perform consistent, atomic snapshots of groups of Volumes using the new Kubernetes [VolumeGroupSnapshot API](https://kubernetes.io/blog/2024/12/18/kubernetes-1-32-volume-group-snapshot-beta/). This capability is especially critical for stateful applications that rely on multiple volumes to ensure data consistency, such as databases and analytics workloads.
+
+## Background
+
+Velero currently enables snapshot-based backups on an individual Volume basis through CSI drivers. However, modern stateful applications often require multiple volumes for data, logs, and backups. This distributed data architecture increases the risk of inconsistencies when volumes are captured individually. Kubernetes has introduced the VolumeGroupSnapshot(VGS) API [(KEP-3476)](https://github.com/kubernetes/enhancements/pull/1551), which allows for the atomic snapshotting of multiple volumes in a coordinated manner. By integrating this feature, Velero can offer enhanced disaster recovery for multi-volume applications, ensuring consistency across all related data.
+
+## Goals
+- Ensure that multiple related volumes are snapshotted simultaneously, preserving consistency for stateful applications via VolumeGroupSnapshots(VGS) API.
+- Integrate VolumeGroupSnapshot functionality into Velero’s existing backup and restore workflows.
+- Allow users to opt in to volume group snapshots via specifying the group label.
+
+## Non-Goals
+- The proposal does not require a complete overhaul of Velero’s CSI integration, it will extend the current mechanism to support group snapshots.
+- No any changes pertaining to execution of Restore Hooks
+
+## High-Level Design
+
+Backup workflow:
+- Accept the label to be used for VGS from the user:
```

Comment:
Is there only one label, or could I have multiple labels?

Should this be an annotation and not a label?

Reply:
There is only one label. Here we are only accepting the label key from the user. The grouping will be done based on the label key and its corresponding value on PVC. For example if the user specifies the label as `vgs-label` and there are v PVCs inthe env; 2 with label `vgs-label=group-1` and other 2 label `vgs-label=group-2` then in such a scenario we would have 2 VGS for each group's PVCs i.e. one VGS for `group-1` and another VGS for `group-2`.

Reply:
@shawn-hurley it's a label (and not an annotation) because that's how VGS works. You create a VGS with a label selector, and it creates a VS for every PVC that matches the label selector.

Reply:
While we could accept multiple labels and then use whichever label is found on the PVC to create the VGS label selector, I'm not sure the complexity is warranted -- but that's an easy future feature if we decide we want to add it. Basically in v2 of this, allow the label to look for to be a comma-separated list instead of a single label string.